src/server/search/aggregator.cc

```diff
@@ -10,62 +10,86 @@ namespace dfly::aggregate {
+  ValuesList out(fields.size());
+  for (size_t i = 0; i < fields.size(); i++) {
+    auto it = dv.find(fields[i]);
+    out[i] = (it != dv.end()) ? it->second : Value{};
+  }
+  return out;
+}
 
-    absl::flat_hash_set<std::string> fields_to_print;
-    fields_to_print.reserve(fields_.size() + reducers_.size());
+DocValues PackFields(ValuesList values, absl::Span<const std::string> fields) {
+  DCHECK_EQ(values.size(), fields.size());
+  DocValues out;
+  for (size_t i = 0; i < fields.size(); i++)
+    out[fields[i]] = std::move(values[i]);
+  return out;
+}
 
-    for (auto& field : fields_) {
-      fields_to_print.insert(std::move(field));
-    }
-    for (auto& reducer : reducers_) {
-      fields_to_print.insert(std::move(reducer.result_field));
-    }
+const Value kEmptyValue = Value{};
+
+}  // namespace
 
-    return {std::move(out), std::move(fields_to_print)};
+void Aggregator::DoGroup(absl::Span<const std::string> fields, absl::Span<const Reducer> reducers) {
+  // Separate items into groups
+  absl::flat_hash_map<ValuesList, std::vector<DocValues>> groups;
+  for (auto& value : result.values) {
+    groups[ExtractFieldsValues(value, fields)].push_back(std::move(value));
   }
 
-  absl::FixedArray<Value> Extract(const DocValues& dv) {
-    absl::FixedArray<Value> out(fields_.size());
-    for (size_t i = 0; i < fields_.size(); i++) {
-      auto it = dv.find(fields_[i]);
-      out[i] = (it != dv.end()) ? it->second : Value{};
+  // Restore DocValues and apply reducers
+  auto& values = result.values;
+  values.clear();
+  values.reserve(groups.size());
+  while (!groups.empty()) {
+    auto node = groups.extract(groups.begin());
+    DocValues doc = PackFields(std::move(node.key()), fields);
+    for (auto& reducer : reducers) {
+      doc[reducer.result_field] = reducer.func({reducer.source_field, node.mapped()});
     }
-    return out;
+    values.push_back(std::move(doc));
   }
 
-  DocValues Unpack(absl::FixedArray<Value>&& values) {
-    DCHECK_EQ(values.size(), fields_.size());
-    DocValues out;
-    for (size_t i = 0; i < fields_.size(); i++)
-      out[fields_[i]] = std::move(values[i]);
-    return out;
+  auto& fields_to_print = result.fields_to_print;
+  fields_to_print.clear();
+  fields_to_print.reserve(fields.size() + reducers.size());
+
+  for (auto& field : fields) {
+    fields_to_print.insert(field);
   }
+  for (auto& reducer : reducers) {
+    fields_to_print.insert(reducer.result_field);
+  }
+}
 
-  std::vector<std::string> fields_;
-  std::vector<Reducer> reducers_;
-};
+void Aggregator::DoSort(std::string_view field, bool descending) {
+  auto comparator = [&](const DocValues& l, const DocValues& r) {
+    auto l_it = l.find(field);
+    auto r_it = r.find(field);
 
-const Value kEmptyValue = Value{};
+    // Handle cases where one of the fields is missing
+    if (l_it == l.end() || r_it == r.end()) {
+      return l_it != l.end() || r_it == r.end();
+    }
```

Comment:
why descending doesn't affect this part of the code

Reply:
This what redis does:
```
1. l_it == l.end() && r_it != r.end()
   asc -> false
   desc -> false
2. l_it !=  l.end() && r_it == r.end()
  asc -> true
  desc -> true
3. l_it ==  l.end() && r_it == r.end()
   asc -> false
  desc -> false
```

Reply:
So descending doesn't affect this part. But I need to fix it to `l_it != l.end() && r_it == r.end()`